libraries/aws/jobs/src/jobs.c

```diff
@@ -0,0 +1,558 @@
+
+        for( api = JobsDescribeSuccess; api < JobsMaxTopic; api++ )
+        {
+            ret = strnnEq( p, length, apiTopic[ api ], apiTopicLength[ api ] );
+
+            if( ret == JobsSuccess )
+            {
+                *outApi = api;
+                *outJobId = jobId;
+                *outJobIdLength = jobIdLength;
+                break;
+            }
+        }
+    }
+
+    return ret;
+}
+
+/**
+ * See jobs.h for docs.
+ *
+ * @brief Output a topic value if a Jobs API topic string is present.
+ */
+JobsStatus_t Jobs_MatchTopic( const char * topic,
+                              size_t length,
+                              const char * thingName,
+                              uint16_t thingNameLength,
+                              JobsTopic_t * outApi,
+                              char ** outJobId,
+                              uint16_t * outJobIdLength )
+{
+    JobsStatus_t ret = JobsBadParameter;
+    JobsTopic_t api = JobsInvalidTopic;
+    char * jobId = NULL;
+    uint16_t jobIdLength = 0U;
+
+    if( ( topic != NULL ) && ( length > 0U ) && checkThingParams() &&
+        ( outApi != NULL ) )
+    {
+        const char * prefix = topic;
+        const char * name = &prefix[ JOBS_API_PREFIX_LENGTH ];
+        const char * bridge = &name[ thingNameLength ];
+
+        ret = JobsNoMatch;
+
+        /* check the shortest match first */
+        if( ( length > JOBS_API_COMMON_LENGTH( thingNameLength ) ) &&
+            ( strnEq( bridge, JOBS_API_BRIDGE, JOBS_API_BRIDGE_LENGTH ) == JobsSuccess ) &&
+            ( strnEq( prefix, JOBS_API_PREFIX, JOBS_API_PREFIX_LENGTH ) == JobsSuccess ) &&
+            ( strnEq( name, thingName, thingNameLength ) == JobsSuccess ) )
+        {
+            const char * tail = &bridge[ JOBS_API_BRIDGE_LENGTH ];
+            size_t tailLength = length - JOBS_API_COMMON_LENGTH( thingNameLength );
+
+            ret = matchApi( tail, tailLength, &api, &jobId, &jobIdLength );
+        }
+    }
+
+    if( outApi != NULL )
+    {
+        *outApi = api;
+    }
+
+    if( outJobId != NULL )
+    {
+        *outJobId = jobId;
+    }
+
+    if( outJobIdLength != NULL )
+    {
+        *outJobIdLength = jobIdLength;
+    }
+
+    return ret;
+}
+
+/**
+ * See jobs.h for docs.
+ *
+ * @brief Populate a topic string for a GetPendingJobExecutions request.
+ */
+JobsStatus_t Jobs_GetPending( char * buffer,
+                              size_t length,
+                              const char * thingName,
+                              uint16_t thingNameLength,
+                              size_t * outLength )
+{
+    JobsStatus_t ret = JobsBadParameter;
+    size_t start = 0U;
+
+    if( checkCommonParams() )
+    {
+        writePreamble( buffer, &start, length, thingName, thingNameLength );
+
+        ret = strnAppend( buffer, &start, length,
+                          JOBS_API_GETPENDING, JOBS_API_GETPENDING_LENGTH );
+
+        if( start == length )
+        {
+            start--;
+        }
```

Comment:
I know it's not possible in this implementation for `start > length`, but would it be safer to just check anyway, in case something in the implementation changes in the future? 
```suggestion
        if( start >= length )
        {
            start = length - 1U;
        }
```

Reply:
changed